File: moedict-android/app/src/main/java/com/yichiuan/moedict/ui/main/DefinitionAdapter.java
package com.yichiuan.moedict.ui.main;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.text.Spannable;
import android.text.SpannableStringBuilder;
import android.text.style.BulletSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.yichiuan.moedict.R;
import com.yichiuan.moedict.common.ui.StaticTextView;
import com.yichiuan.moedict.common.TextUtil;

import java.util.ArrayList;

import moe.Definition;
import moe.Heteronym;


public class DefinitionAdapter extends RecyclerView.Adapter<DefinitionAdapter.DefinitionViewHolder> {

    static class Section {
        public String type;
        public ArrayList<Definition> definitions = new ArrayList<>();
        public int size() {
            return type == null ? definitions.size() : definitions.size() + 1;
        }
    }

    LayoutInflater inflater;
    Heteronym heteronym;

    ArrayList<Section> sections = new ArrayList<>();
    int sectionSize;

    public DefinitionAdapter(Context context, Heteronym heteronym) {
        inflater = LayoutInflater.from(context);
        this.heteronym = heteronym;
        initSections();
    }

    // Group definitions with the same type into section
    private void initSections() {
        sections.clear();
        sectionSize = 0;

        int definitionLength = heteronym.definitionsLength();

        String currentType = "";
        Section section = null;

        for (int i = 0; i < definitionLength; ++i) {
            Definition definition = heteronym.definitions(i);
            String type = definition.type();

            if (i == 0 ||
                currentType != null && !currentType.equals(type) ||
                currentType == null && type != null) {

                currentType = type;

                section = new Section();
                section.type = type;
                sections.add(section);
                sectionSize++;
            }
            section.definitions.add(definition);
        }
    }

    @Override
    public DefinitionViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new DefinitionViewHolder(inflater.inflate(R.layout.item_definition, parent, false));
    }

    @Override
    public void onBindViewHolder(DefinitionViewHolder holder, int position) {

        Section section = sections.get(position);

        if (section.type == null || section.type.isEmpty()) {
            holder.type.setVisibility(View.GONE);
        } else {
            holder.type.setVisibility(View.VISIBLE);
            holder.type.setText(section.type);
        }

        holder.type.setText(section.type);

        SpannableStringBuilder builder = new SpannableStringBuilder();

        int bulletStart = 0;

        for (int i = 0; i < section.definitions.size(); ++i) {

            Definition definition = section.definitions.get(i);

            builder.append(TextUtil.INSTANCE.obtainSpannedFrom(definition.defAsByteBuffer()));

            for (int j = 0; j < definition.examplesLength(); ++j) {
                if (j != 0 || builder.length() != 0) {
                    builder.append("\n");
                }
                builder.append(TextUtil.INSTANCE.obtainSpannedFrom(
                        definition.examples(j).u16strAsByteBuffer()));
            }

            for (int j = 0; j < definition.quotesLength(); ++j) {
                if (j != 0 || builder.length() != 0) {
                    builder.append("\n");
                }
                builder.append(TextUtil.INSTANCE.obtainSpannedFrom(
                        definition.quotes(j).u16strAsByteBuffer()));
            }

            for (int j = 0; j < definition.linksLength(); ++j) {
                if (j != 0 || builder.length() != 0) {
                    builder.append("\n");
                }
                builder.append(TextUtil.INSTANCE.obtainSpannedFrom(
                        definition.links(j).u16strAsByteBuffer()));
            }

            if (section.definitions.size() > 1) {
                BulletSpan bullet = new BulletSpan();
                builder.setSpan(bullet, bulletStart, builder.length(), Spannable.SPAN_INCLUSIVE_EXCLUSIVE);
            }

            if (i + 1 < section.definitions.size()) {
                builder.append("\n");
                bulletStart = builder.length();
            }
        }

        holder.definition.setText(builder);
    }

    @Override
    public int getItemCount() {
        return sections.size();
    }

    static class DefinitionViewHolder extends RecyclerView.ViewHolder {

        TextView type;
        StaticTextView definition;

        public DefinitionViewHolder(View itemView) {
            super(itemView);
            type = itemView.findViewById(R.id.textview_type);
            definition = itemView.findViewById(R.id.staticview_definition);
        }
    }

}

File: moedict-android/app/src/main/java/com/yichiuan/moedict/ui/main/DefinitionAdapter.kt
package com.yichiuan.moedict.ui.main

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.text.Spannable
import android.text.SpannableStringBuilder
import android.text.style.BulletSpan
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.yichiuan.moedict.R
import com.yichiuan.moedict.common.TextUtil
import com.yichiuan.moedict.common.ui.StaticTextView
import moe.Definition
import moe.Heteronym
import java.util.*


class DefinitionAdapter(context: Context, private var heteronym: Heteronym
) : RecyclerView.Adapter<DefinitionAdapter.DefinitionViewHolder>() {

    private var inflater: LayoutInflater = LayoutInflater.from(context)

    private var sections = ArrayList<Section>()

    private var sectionSize: Int = 0

    class Section {
        var type: String? = null
        var definitions = ArrayList<Definition>()
        fun size(): Int {
            return if (type == null) definitions.size else definitions.size + 1
        }
    }

    init {
        initSections()
    }

    // Group definitions with the same type into section
    private fun initSections() {
        sections.clear()
        sectionSize = 0

        val definitionLength = heteronym.definitionsLength()

        var currentType: String? = ""
        var section: Section? = null

        for (i in 0 until definitionLength) {
            val definition = heteronym.definitions(i)
            val type = definition.type()

            if (i == 0 ||
                    currentType != null && currentType != type ||
                    currentType == null && type != null) {

                currentType = type

                section = Section()
                section.type = type
                sections.add(section)
                sectionSize++
            }
            section!!.definitions.add(definition)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DefinitionViewHolder {
        return DefinitionViewHolder(inflater.inflate(R.layout.item_definition, parent, false))
    }

    override fun onBindViewHolder(holder: DefinitionViewHolder, position: Int) {

        val section = sections[position]

        if (section.type.isNullOrEmpty()) {
            holder.type.visibility = View.GONE
        } else {
            holder.type.visibility = View.VISIBLE
            holder.type.text = section.type
        }

        holder.type.text = section.type

        val builder = SpannableStringBuilder()

        var bulletStart = 0

        for (i in section.definitions.indices) {

            val definition = section.definitions[i]

            builder.append(TextUtil.obtainSpannedFrom(definition.defAsByteBuffer()))

            for (j in 0 until definition.examplesLength()) {
                if (j != 0 || builder.isNotEmpty()) {
                    builder.append("\n")
                }
                builder.append(TextUtil.obtainSpannedFrom(
                        definition.examples(j).u16strAsByteBuffer()))
            }

            for (j in 0 until definition.quotesLength()) {
                if (j != 0 || builder.isNotEmpty()) {
                    builder.append("\n")
                }
                builder.append(TextUtil.obtainSpannedFrom(
                        definition.quotes(j).u16strAsByteBuffer()))
            }

            for (j in 0 until definition.linksLength()) {
                if (j != 0 || builder.isNotEmpty()) {
                    builder.append("\n")
                }
                builder.append(TextUtil.obtainSpannedFrom(
                        definition.links(j).u16strAsByteBuffer()))
            }

            if (section.definitions.size > 1) {
                val bullet = BulletSpan()
                builder.setSpan(bullet, bulletStart, builder.length, Spannable.SPAN_INCLUSIVE_EXCLUSIVE)
            }

            if (i + 1 < section.definitions.size) {
                builder.append("\n")
                bulletStart = builder.length
            }
        }

        holder.definition.setText(builder)
    }

    override fun getItemCount(): Int {
        return sections.size
    }

    class DefinitionViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        var type: TextView = itemView.findViewById(R.id.textview_type)
        var definition: StaticTextView = itemView.findViewById(R.id.staticview_definition)
    }
}
